analyzers/src/SonarAnalyzer.Common/AnalysisContext/SonarReportingContextBase.cs

```diff
@@ -0,0 +1,55 @@
+﻿/*
+ * SonarAnalyzer for .NET
+ * Copyright (C) 2015-2022 SonarSource SA
+ * mailto: contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+namespace SonarAnalyzer;
+
+public abstract class SonarReportingContextBase<TContext> : SonarAnalysisContextBase<TContext>
+{
+    private protected abstract ReportingContext CreateReportingContext(Diagnostic diagnostic);
+
+    protected SonarReportingContextBase(SonarAnalysisContext analysisContext, TContext context) : base(analysisContext, context) { }
+
+    public void ReportIssue(Diagnostic diagnostic)  // FIXME: Make this obsolete
```

Comment:
I would change the method name to `Report`. Roslyn uses the term `Diagnostic` not `Issue` and that makes the name confusing for me. I wouldn't add `Diagnostic` to the name since it's already provided as a parameter.

Reply:
I like the idea. I will add it to the `FIXME` part, as renaming this one would change all rules. So the <Obsolete> will be a good opportunity to do it in a clean way

Reply:
Actually, the new methods will be called like
```
context.ReportIssue(Rule, c.Node.GetLocation()));
```
with signature like this (and similar)
```
ReportIssue(DiagnosticDescriptor descriptor, Location location)
```
Should it be still only `Report` when it doesn't receive diagnostic itself but will create one?

Reply:
No, in this case we can keep the name.